### ISSUE TITLE
Support for NVENC (nvEncodeAPI64.dll)

### ISSUE THREAD

**Issue by Nik77x** [SveSop/nvidia-libs#25]
I was trying to run an app using the latest version of wine-staging (9.21) through bottles. The app required nvcuda.dll and nvEncodeAPI64.dll, I was able to add nvcuda from the files provided in the release, but I couldn't find nvEncodeAPI64.dll anywhere. I searched around, and the last version of wine that supported cuda and nvenc is 7.19, but it doesn't work with the app I'm trying to run, and copying the .dll from that version obviously didn't work. I was wondering if there was any way I could get nvenc working, or a workaround to make the app work without it.

If that's useful the app in question is [Topaz VideoAI](https://www.topazlabs.com/topaz-video-ai) (testing with the trial version) and when rendering the preview it fails with "Driver error" and these are the logs:

```
2024-11-15 19-44-31.938 Thread: 968 Info FF Process Output: 57 [48;5;0m[38;5;0m[h264_nvenc @ 0000000001A90840] [0m[48;5;0m[38;5;0mCannot load nvEncodeAPI64.dll

[0m[48;5;0m[38;5;0m[h264_nvenc @ 0000000001A90840] [0m[48;5;0m[38;5;0mThe minimum required Nvidia driver for nvenc is 531.61 or newer

[0m[48;5;0m[38;5;0m[vost#0:0/h264_nvenc @ 0000000001A905C0] [0m[48;5;0m[38;5;0mError while opening encoder - maybe incorrect parameters such as bit_rate, rate, width or height.

[0m[48;5;0m[38;5;0m[fc#0 @ 0000000001A7C080] [0m[48;5;0m[38;5;0mError sending frames to consumers: Operation not permitted
   
```

**Comment by SveSop** [comment]
NVENC has not been under development for many many years in the old staging tree (wine 7.xx), so i do not find it strange that this is not supported anymore.

I am somewhat reluctant to take on nvenc, as this is not documented as well as nvcuda. nvcuda is currently living on its last leg when it comes to wine i suppose, as i may be inclined to think that the shift to wow64 in wine may make this stop working altogether at some point.

The undertaking of re-writing nvcuda to support the "wow64 method" is somewhat beyond my skills, and if nvenc should be added at THIS time, it would be futile to make it work the same way as nvcuda currently does... as that would add yet another hurdle of rewriting in the future.

So, as of now - it is a "wontfix" from me sadly.

**Comment by Nik77x** [comment]
Ok, that's what I thought. Doesn't matter, I guess I'll stick to dual booting. 
I hope eventually there will be support for cuda and nvenc again in wine, it's quite important for a number of art/editing software.
Thanks for the help.

**Comment by SveSop** [comment]
Do you have an example of something freeware that extensively uses nvenc? (As i am not bying software for $299 just for testing).

Afaik wine tries to have a more independent view of vendors and work mostly with opensource. Binaries and drivers++ tied to particular vendor like NVIDIA has less chance of being adopted in wine "upstream", so the "staging" version was more or less dropped developing as it had little chance of actually getting upstreamed, and i do not really expect this to change.

CUDA should be supported up to CUDA 12.6 with nvidia-libs as long as you use a driver that also support this ofc. Do not mix cuda/nvenc, as these serve different purposes 😄

**Comment by Nik77x** [comment]
> Do you have an example of something freeware that extensively uses nvenc?

The simplest example that I can think of is [handbrake](https://github.com/HandBrake/HandBrake), it already has a linux version but for testing in wine it could work.

If you want to test with my same setup, you don't need to buy VideoAi, as they have a fully-featured trial version with exporting locked. That's what I was using to see if I could get it to work.

> CUDA should be supported up to CUDA 12.6 with nvidia-libs as long as you use a driver that also support this ofc. Do not mix cuda/nvenc, as these serve different purposes

Yeah, with that I meant directly into wine without using nvidia-libs, but I understand why upstream decided to get rid of it. Nvidia seems really hard to work with thanks to all their proprietary things.

**Comment by SveSop** [comment]
I am going to take a mild stab at this, but make no promises, and my skills are limited, so it will be somewhat of the same solution as nvcuda is atm - A linux library disguised as a windows .dll.

Keeping this open, and see if i can get handbrake to recognize that nvencodeapi is available.

**Comment by SveSop** [comment]
I pushed a standalone version with some updates for SDK 10.0 here: https://github.com/SveSop/nvencodeapi/tree/devel

I tested this with HandBrake, and it did indeed seem to use around 45% "ENC" usage, and 20-25% GPU usage according to nvtop when using it with a HQ profile and H265(nvenc) codec.

[nvencodeapi.zip](https://github.com/user-attachments/files/17791891/nvencodeapi.zip)

In case you don't compile yourself.

For now, manual install.. You can symlink or copy `nvencodeapi64.dll.so` as `windows\system32\nvencodeapi64.dll` in your wineprefix. (and similar for nvencodeapi.dll.so in the syswow64 windows folder in case of 32-bit).

**Comment by Nik77x** [comment]
I tried this with both handbrake and videoAi, but it doesn't seem to be detected at all. Handbrake doesn't enable support for nvenc encoding (in preferences → video) and videoAi gives me the same error as before (nvEncodeAPI64.dll not found). I don't know if I'm doing something wrong, which would be weird since I did exactly the same thing with nvcuda and it works.

What I did was:

1) Copy the nvencodeapi64.dll.so to windows\system32\nvencodeapi64.dll 
     1.1) Make a copy calling it nvEncodeAPI64.dll for good measure
2) Copy nvencodeapi.dll.so to windows\syswow64\nvencodeapi.dll
3) Set dll overrides to native for all of them in bottles.

The same thing I did for nvcuda.dll, and it works, tested also the dll override syntax, and it's right. I really don't know what's going on...

**Comment by SveSop** [comment]
Been having some fun fiddling with this, and found that i could use MPC-HC (media player).. an oldie but goodie, and by incorporating `nvcuvid.dll` it plays videos with HW decoding too.

Now, i also found out that using HandBrake, there is an option when generating previous and stuff, it will also use `nvcuvid` for some hardware decoding (given you select the option to do so i guess).

So, since this is more "NVENC" than just nvencodeapi, i renamed the repo, and added both here: https://github.com/SveSop/nvenc/tree/devel

Sorry for the inconvenience in case you cloned this tho.. Here is a binary:
[nvenc.zip](https://github.com/user-attachments/files/17819454/nvenc.zip)

Same deal here for the `nvcuvid.dll.so` located in the lib/lib64 folder structure. Maybe Topaz VideoAI also tries to load nvcuvid for something? Dunno.. 

Will see if i can figure out how to get this working in a bottle with said HandBrake and MPC-HC, and let you know.

**Comment by Nik77x** [comment]
I kept testing with handbrake to be sure and ENC in nvtop goes up to 100%, so it works properly.
I re-ran the encoding a few times to see if it worked reliably, and it works until it says: 
`[h264_nvenc @ 00007adac912b140] dl_fn->cuda_dl->cuCtxCreate(&ctx->cu_context_internal, 0, cu_device) failed -> CUDA_ERROR_OUT_OF_MEMORY: out of memory
[h264_nvenc @ 00007adac912b140] No capable devices found`
which is weird since in nvtop and system monitor the vram is half empty and doesn't go up when starting encoding. But most of the time it works.
I also tested all the nvenc options (h.264, h.265 and h.265 10 bit) and all of them work with full utilization

> Sorry for the inconvenience in case you cloned this tho..
No problem, I had just downloaded the binary, and it wouldn't have been a problem anyway ;)

**Comment by Nik77x** [comment]
Ok, I tested the script, and it works great! That's a huge time saver 👍

For testing with topaz, I tried installing it again in a new bottle, and it gave me the same error you described, then I just switched runners to kron4ek-wine-9.21-staging-amd64, ran it, then switched it back to kron4ek-wine-9.21-amd64 and it worked. No additional dependencies added.
This is the bottle configuration:

[TopazVideoAI_Conf.zip](https://github.com/user-attachments/files/17868382/TopazVideoAI_Conf.zip)

I also updated VideoAI to [5.4](https://community.topazlabs.com/t/topaz-video-ai-5-4/81539), they swapped the video player with [mpv](https://github.com/mpv-player/mpv), so I was able to test hardware decoding and it works well. Now video playback flickers but that's another problem...

Thanks again for all the awesome work you're doing! 🙂

**Comment by SveSop** [comment]
Luckily you can still run 64-bit with the Kron4tek builds, but afaik this is a wow64 build that would completely bork 32-bit library usage... just so you are aware.
https://github.com/SveSop/nvidia-libs/issues/25#issuecomment-2480588866

Using your settings it seemed to work.. weirdly enough.. Maybe something to do with "use decorations" disabled, as this is normally enabled. Not sure. Other libs and wine versions were the same.

The flickering from the video player is annoying. Not sure what that would be, but it sure does use GPU 👍 VideoAI is at 5.5 now tho lol.. so atleast they seem to be updating. Maybe ill see if i can figure out something with mpv, but could be wine related.

**Comment by SveSop** [comment]
Have been fiddling a bit with this to try to find out why the stupid flashing happens. So, i scratched and re-did the bottle many times, and there is definately something very strange going on if you are using the default DPI of 96. The app does not launch at all for some reason if you do not set something like 100 or whatever. If it is set at 96, it just ends.

Now, there are some thingys in wine that have to do with DPI scaling and whatnot, so it could be something going on there. I tried this with the GE-Proton build, but there it did not start AT ALL, so Proton might have patched away some of wine upstream dpi scaling stuff... Just found that somewhat interesting. 

Another thing to note is that with this new MPV player, there is also some codec related dll's missing. avcodec-60.dll and whatnot. This was not there with version 5.3, so i guess this is something new related to FFMpeg usage or this MPV player. Possibly it is meant to be installed if the damn installer had worked as it should have done, but .msi installers is not really wine's point of expertise i suppose.

This brings us a bit back to MPV. Running this player as a standalone player FROM the Topaz folder seems to play videos just fine without flickering, so i do not think THAT is the source of the issues. It is most likely some crud with the Topaz main window that is causing all kinds of crap (like the one with DPI scaling). If you use the "scrubber" (timeline thing) and move it to fast forward, but do NOT move the mouse pointer, the image is shown, but immediately upon moving the mouse -> WHAM black. 

Same if you make the window smaller and keep moving it around the desktop WHILE playing a video, it seems to play fairly oki.. but if you slow down the movements or stop, the flickering increases. I think this is some overlay feature thing that is not supported too well in wine, so i doubt there is anything wrong with NVENC or NVCUDA implementation that causes this... atleast not that i can understand.

The 5.3.6 version seems to work just fine, although it does not seem to use NVCUVID for hardware decoding when playing like 5.4 and 5.5 version does with MPV player. It should be no huge issue i suppose? Atleast until that video flickering can be figured out. I will keep looking into this, and see if it is the program itself or someting very strange happening when nvcuvid sends data to nvcuda maybe.

Only thing that seems an "requirement" to even run is a DPI scale of >96. Very strange indeed. Could be some dependency would fix that.

**Comment by Nik77x** [comment]
Thanks for looking into it! Looking at it quickly, the preview becomes black just when the "unprocessed video" overlay comes up (when scrubbing it doesn't show it) and it doesn't seem to be a wine related issue since [here](https://community.topazlabs.com/t/topaz-video-ai-5-4/81539/16) there's a guy having the same problem, and it's highly unlikely he's using linux... and he also says he has the same problem with the previews for all videos, so it does seem to be the same problem. I think it may be something Topaz has to fix, and nothing can be done user-side.

> The 5.3.6 version seems to work just fine, although it does not seem to use NVCUVID for hardware decoding

All versions before 5.4 weren't hardware accelerated so it makes sense that they wouldn't use NVCUVID, they replaced the player with mpv for this reason I think.

> Only thing that seems an "requirement" to even run is a DPI scale of >96.

That's just what running apps with wine is like lol. It's an art, the weirdest solution is usually the correct one. I didn't even know that's what made it work... I just set the dpi to 130 to all apps because I'm on a 1440p screen... 😂

Anyway, for now it's good that it works at all, I'm Ok waiting for them to fix their bugs, 5.3 works fine, for everything else dual booting also works.

About the installer, after running it crashes with (something)CachePackage script failed to execute, I tried installing dependencies loosely related to scripts, like powershell core and wrapper, dmscript and whscript57 (I don't know what they do exactly, but I wanted to try anyway) but it still didn't work. After that I stopped looking into as I saw that it still installed the required files to run and killing the process worked and just moved on.

**Comment by SveSop** [comment]
I made a new release with some tweaks and updates, but not something that address this directly.

I also tested a bit back and forth trying various solutions, but cant really come up with anything better than some Qt6 stuff not being overly friendly when running under Wine tbh. Wine is a bit behind on that stuff, as i think it is not too long ago Qt5 apps had issues there, so maybe some day in the future it will be better.

Ie. conclusion from my side would be some overlay stuff when using Qt6 that may not play too nice with wine.

So, for the issue at hand, it does indeed seem to work now, so going to close this. Feel free to open a new issue in case there is something related to nvenc or nvcuda 😄
